### PR TITLE
New style GetItem

### PR DESCRIPTION
**Pull request by delta2323**
This PR implements new-style version of `F.get_item` as a part of #3147. This function will be used to make `F.softmax_cross_entropy` double-backpropable.

### CODE REVIEW

File: chainer/functions/array/get_item.py [hunk 1]
```diff
@@ -37,22 +37,35 @@ def check_type_forward(self, in_types):
...
         self._in_shape = ary.shape
         self._in_dtype = ary.dtype
         return utils.force_array(ary[self.slices]),
 
-    def backward(self, xs, gys):
-        xp = cuda.get_array_module(*gys)
-        gy = gys[0]
+    def backward(self, indexes, gy):
+        return GetItemGrad(
+            self.slices, self._in_shape, self._in_dtype).apply(gy)
```
Comment (beam2d) [inline chainer/functions/array/get_item.py]: Input shape and dtype can be obtained by `self.inputs[0].{shape,dtype}`, so you do not have to save them in forward.

**Comment by delta2323** [comment]
Thank you for your comments. I updated the code.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
Jenkins, test this please.

**Comment by beam2d** [comment]
Jenkins, test this please

**Comment by beam2d** [comment]
LGTM

**Comment by beam2d** [comment]
Thanks!

**Pull request merged**